### PR TITLE
AirfRANS: retest 4L/256d capacity with no-EMA + single jobs at slices=96

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #2429 tested 4L/256d capacity but with epoch starvation (only 2 epochs at slices=64 with 4 parallel jobs). The capacity hypothesis is unresolved. Now, with no-EMA proven critical (PR #2412/#2431) and understanding that parallel jobs kill epoch count, we retest 4L/256d with proper controls: single jobs at slices=96 for fair epoch count comparison.

**Critical instruction:** Run as SINGLE JOBS on separate GPUs. Do NOT run 4 parallel jobs — that caused epoch starvation in Round 2.

**Critical config change:** Use `--no-use-ema`. This is now mandatory.

## Current Baseline

- **AirfRANS val_primary/surface_mse: 0.3308** (PR #2423 — 3L/192d, AdamW lr=5e-4, EMA=True, slices=96, 6 epochs)
- **AirfRANS test_primary/surface_mse: 0.3199**

## Instructions

Run 2 trials on separate GPUs. Use `--wandb_group radford-airfrans-noema` for all runs.

### Trial 0 — GPU 0: 4L/256d + no-EMA + AdamW lr=5e-4 at slices=96 (SINGLE JOB)
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 96 --wandb_group radford-airfrans-noema --wandb_name emma/airfrans-noema-4l256d-adamw-5e4 --agent emma
```

### Trial 1 — GPU 1: 3L/192d + no-EMA + AdamW lr=5e-4 at slices=96 (matched control)
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --wandb_group radford-airfrans-noema --wandb_name emma/airfrans-noema-3l192d-adamw-5e4 --agent emma
```

Run both simultaneously (`&`). Each should complete 5-6 epochs in 30 minutes (4L/256d may be slightly slower).

**Key question:** Does 4L/256d outperform 3L/192d when properly trained (6 epochs, no-EMA)? Report:
- val_primary/surface_mse (best across epochs) for each trial
- Epochs completed per trial (this verifies no starvation)
- surface_mse_p breakdown

## Results

_To be filled by student after runs complete._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT emma:

## Results

### Trial 0 — 4L/256d + no-EMA + AdamW lr=5e-4 at slices=96 (GPU 0)

| Epoch | val_primary/surface_mse | loss |
|-------|------------------------|------|
| 1 | 0.5172 | 1.573 |
| 2 | 0.4107 | 1.367 |
| 3 | 0.3338 | 1.250 |
| 4 | 0.2954 | 1.149 |
| **5** | **0.2935** | 1.132 |

**Test metrics (best model at epoch 5):**
- test_primary/surface_mse: **0.2706**
- surface_mse_p: 1.0808
- surface_mse_Ux: 0.00133
- surface_mse_Uy: 0.00016
- surface_mse_nut: 7e-05

**Epochs completed:** 5 (timeout at 30 min)
**Peak VRAM:** ~51 GB (GPU 0)
**W&B run:** `wptz6kat`
**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 96 --epochs 30 --wandb-group radford-airfrans-noema --wandb-name "emma/airfrans-noema-4l256d-adamw-5e4-v2" --agent emma
```

---

### Trial 1 — 3L/192d + no-EMA + AdamW lr=5e-4 at slices=96 (GPU 2)

| Epoch | val_primary/surface_mse | loss |
|-------|------------------------|------|
| 1 | 0.4637 | 1.564 |
| 2 | 0.3947 | 1.356 |
| 3 | 0.2968 | 1.216 |
| 4 | 0.3090 | 1.174 |
| 5 | 0.2673 | 1.133 |
| **6** | **0.2597** | 1.050 |

**Test metrics (best model at epoch 6):**
- test_primary/surface_mse: **0.2392**
- surface_mse_p: 0.9556
- surface_mse_Ux: 0.00121
- surface_mse_Uy: 4e-05
- surface_mse_nut: 2e-05

**Epochs completed:** 6 (timeout at 30 min)
**Peak VRAM:** ~57 GB (GPU 2)
**W&B run:** `pifi0x1v`
**Command:**
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --epochs 30 --wandb-group radford-airfrans-noema --wandb-name "emma/airfrans-noema-3l192d-adamw-5e4-v3" --agent emma
```

---

### Comparison vs Baseline

| Config | Epochs | val_primary/surface_mse | test_primary/surface_mse | surface_mse_p (test) |
|--------|--------|------------------------|-------------------------|---------------------|
| **Baseline** (3L/192d, EMA=True) | 6 | 0.3308 | 0.3199 | — |
| **4L/256d, no-EMA** | 5 | 0.2935 (−11.3%) | 0.2706 (−15.4%) | 1.0808 |
| **3L/192d, no-EMA** | 6 | 0.2597 (−21.5%) | 0.2392 (−25.2%) | 0.9556 |

### What happened

**The 3L/192d model outperforms 4L/256d** — the opposite of the capacity hypothesis. Key observations:

1. **3L/192d is clearly better:** val 0.2597 vs 0.2935, test 0.2392 vs 0.2706. The smaller model wins by a large margin on every metric.
2. **Epoch advantage matters:** 3L/192d completed 6 epochs vs 4L/256d's 5 (larger model is ~20% slower per epoch). Under a fixed time budget, the smaller model gets more training iterations.
3. **No-EMA is the big win:** Both no-EMA runs massively beat the EMA baseline. The 3L/192d no-EMA run improved test surface_mse by 25.2% over the EMA baseline.
4. **Pressure dominates:** surface_mse_p is >99% of total surface_mse in both runs — pressure prediction remains the bottleneck.
5. **Still improving at timeout:** Both models show monotonically decreasing loss and mostly decreasing val_mse through their final epochs, suggesting more epochs would help.

**Answer to the key question:** 4L/256d does NOT outperform 3L/192d when properly trained. The larger model is both slower (fewer epochs per time budget) and converges to worse metrics. The capacity increase from Round 2 was not needed — the real fix was removing EMA.

### Note on execution

The PR commands omitted `--epochs`, which defaults to 2 in the config (SENPAI_MAX_EPOCHS caps but doesn't raise). I added `--epochs 30` to both commands to allow proper training within the 30-min timeout. Also, Trial 1 got stuck on GPU 1 during the second attempt and had to be relaunched on GPU 2 (v3 run).

### Suggested follow-ups

- **More epochs:** Both models are still improving — try longer training (e.g. 60-min timeout or higher epoch count) to see where 3L/192d plateaus
- **Pressure-focused loss:** Since surface_mse_p dominates, a pressure-weighted loss could accelerate convergence
- **Learning rate tuning with no-EMA:** The 5e-4 LR was tuned with EMA — the no-EMA regime might benefit from a different LR

**Comment by tcapelle** [comment]
ADVISOR: Outstanding result emma! **3L/192d WITHOUT Fourier at 6 epochs = val 0.2597** — beats the Fourier baseline (0.2891) by 10.2%.

CRITICAL FINDING: Fourier features add ~3x epoch overhead on AirfRANS (15 min/epoch vs 5 min/epoch). Without Fourier, you get 6 epochs → 0.2597. With Fourier, only 2 epochs → 0.2710 at best. **More epochs beats Fourier features**, same pattern as TandemFoil slices=64 breakthrough.

4L/256d is conclusively worse than 3L/192d under time budget (fewer epochs, no quality gain).

Merging as new AirfRANS baseline. The no-Fourier + 6 epochs recipe is now the AirfRANS golden config.
